python/pyarrow/_compute.pyx

```diff
@@ -1322,7 +1322,39 @@ class MakeStructOptions(_MakeStructOptions):
 
 cdef class _StructFieldOptions(FunctionOptions):
     def _set_options(self, indices):
-        self.wrapped.reset(new CStructFieldOptions(indices))
+        cdef:
+            CFieldRef field_ref
+            const CFieldRef* field_ref_ptr
+
+        # List[str]/List[bytes] converted to '.a.dotted.path'
+        if isinstance(indices, list) and len(indices):
+            if isinstance(indices[0], str):
+                indices = '.' + '.'.join(indices)
+            elif isinstance(indices[0], bytes):
+                indices = b'.' + b'.'.join(indices)
```

Comment:
This disallows doing a mix of string and int? Alternatively, I would maybe avoid converting a list to a DottedPath, but rather add a generic list handling case. 

More in general, I think we should align this more with the existing method to create a FieldRef in Python/cython (eg `FieldRef._nested_field` constructor). If we have a list input here, we can just pass it through, and then handle the resulting Expression as you already do here.

Reply:
That's good! Updated in https://github.com/apache/arrow/pull/14495/commits/b3219ff46563a22af2e7449569e6172ba70c875c

Reply:
Thanks! Can you also add a test for the changed `pc.field(..)` behaviour itself? (without using the struct_field kernel)

Reply:
https://github.com/apache/arrow/pull/14495/commits/8fbe3380f546d33273d17a73e2bc133d09b6f38b